### PR TITLE
Migrate BAL from main OpenAssetIO repo

### PR DESCRIPTION
**Pull request by foundrytom**
Imports the existing BAL plugin from the main OpenAssetIO repo, preserving history.

Several minor tweaks to cover the differences in deployment/testing now it is a stand-alone repository starting [here](https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/commit/5fe6237ba0a363d5c4f6f2480d1867671544ec6f).

Will set up CI once we have `openassetio` in PyPI

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,6 +28,8 @@ manager behavior.
 - If no `library_path` has been specified, the `BAL_LIBRARY_PATH` env
   var will be checked to see if it points to a valid library file.
 
+- Persists newly registered data in-memory.
```
Comment (elliotcmorris) [inline README.md:31]: [consider][minor] Seems a bit vague to be a feature. Was the motivation behind this this actually to imply that newly registered data won't get serialized back to a `.json`?

**Review by elliotcmorris — APPROVED** [review]
Only reviewed the new changes file by file. Nice work keeping all the history preserved.

File: README.md (hunk 1 above)
Comment (foundrytom) [reply README.md:31]: Exactly that. It wasn't capable of registering new data before (read-only) so it was something we explicitly added. I can elaborate a little more if you think its useful.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+## Testing
+
+The test fixtures take care of configuring the OpenAssetIO plugin search
+paths for you. Assuming your working directory is set to the root of
+this checkout:
+
+```bash
+python -m venv .venv
+. .venv/bin/activate
+python -m pip install -r tests/requirements.txt
```
Comment (feltech) [inline README.md:61]: [consider] Tests don't work unless `openassetio` is available and is built against the same version of Python.

Perhaps worth adding a note to either set `PYTHONPATH` or `pip install openassetio` (a bit pre-emptive, but we're nearly there!).  I understand this should be clear from the "The plugin requires `openassetio` to be available" note above, but I think people will try to run the snippet, perhaps thinking the `requirements.txt` includes `openassetio`.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+- Resolves references with the `bal:///` prefix to data from a
+  pre-configured library of assets stored in a `.json` file.
+
+- The library file to be used is controlled by the `library_path`
+  setting, and this should point to a library file with valid content.
+
+- If no `library_path` has been specified, the `BAL_LIBRARY_PATH` env
+  var will be checked to see if it points to a valid library file.
+
+- Persists newly registered data in-memory.
```
Comment (feltech) [inline README.md:31]: [consider][minor] Perhaps elaborate slightly
```suggestion
- Persists newly registered data in-memory (only).
```

**Review by feltech — APPROVED** [review]
Looks good to me, very minor considerations on the README, otherwise good to go I think.

File: README.md (hunk 2 above)
Comment (foundrytom) [reply README.md:61]: I was going to update all this once we have PyPI in place and it is all a lot more straight forward...

File: README.md (hunk 3 above)
Comment (foundrytom) [reply README.md]: Hows [this](https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/1/commits/f2b3405d15e3e89527c08c4964f26da8abc5a3cf).

Comment (feltech) [reply README.md]: Perfect, thanks

**Pull request merged**